### PR TITLE
Add Brier Score metrics

### PR DESCRIPTION
**Pull request by Vincent-Maladiere**
**What does this PR implement?**

This is the implementation of the Brier score metrics module:
- `BrierScoreComputer`
- <del>`BrierScoreSampler`</del> to be moved to the GBIncidence source code
- `brier_score`
- `brier_score_incidence`
- `integrated_brier_score`
- `integrated_brier_score_incidence`

This also implements the `IPCWEstimator` used by the BrierScoreComputer. We have a dependency on lifelines but not on scikit-survival.

**Remarks**

~~- This fixes a subtle bug in `y_binary`, making our previous implementation of `brier_score()` incorrect. It now matches scikit-survival results.~~
~~**Edit**: `y_true_binary` is 1 when we observed an event, and 0 otherwise, which is correct for our Gradient Boosting **classification or regression target**.
However, when computing the **Brier Score metric**, we actually need `np.logical_not(y_true_binary)`. Indeed, when we observed an event, the survival probability drops to 0. Otherwise, it stays at 1. The Brier Score uses the survival probability, not the incidence probability.~~

- `brier_score` computes is the regular Brier Score like scikit-survival does, where the `y_pred` input is the survival probability. On the contrary, `brier_score_incidence` follows the Kretowska formula for competing events and expects `y_pred` to be the incidence probability for the kth cause of failure instead. Under the hood though, `brier_score` relies on `brier_score_incidence`, by simply inputting `1 - y_pred`.

	This is needed to disambiguate our Brier Score implementation. We could also only expose `brier_score_incidence` without the regular `brier_score`, but I thought it was handy to have it, instead of telling the user to perform `1 - y_pred` for binary events, all the time.

- When running
  ```python
  brier_score(y_train, y_test, y_pred, times)
  ```
  `times` must be the times used for computing `y_pred` —of shape `(n_samples, n_times)`— otherwise, the Brier Score will be incorrect. This is tricky to check when `y_pred` is not a dataframe, so we only check for shapes.

### CODE REVIEW

File: hazardous/metrics/brier_score.py [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+
+        """
+        if (self.event_ids_ > 0).sum() > 1 and self.event_of_interest != "any":
+            warnings.warn(
+                "Computing the Brier Score only make "
+                "sense when the model is trained with a binary event "
+                "indicator or when setting event_of_interest='any'. "
+                "Instead this model was fit on data with event ids "
+                f"{self.event_ids_.tolist()} and with "
+                f"event_of_interest={self.event_of_interest}."
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:85]:
```suggestion
                "Instead this model is evaluated on data with event ids "
                f"{self.event_ids_.tolist()} and with "
                f"event_of_interest={self.event_of_interest}."
```

File: hazardous/metrics/brier_score.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+                event=self.any_event_train,
+                duration=self.duration_train,
+            )
+        )
+
+        # Precompute the censoring probabilities at the time of the events on the
+        # training set:
+        self.ipcw_train = self.ipcw_est.predict(self.duration_train)
+
+    def brier_score(self, y_true, y_pred, times):
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:53]: I would name this `brier_score_survival` to make it explicit that we expect `y_pred` to hold `1 - incidence_probability`.

File: hazardous/tests/test_metrics.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            1.0141844,
+            1.05622555,
+            1.02340426,
+            1.0892326,
+            1.0141844,
+            1.05622555,
+            1.0892326,
+            1.03891038,
+        ]
+    )
```
Comment (ogrisel) [inline hazardous/tests/test_metrics.py:197]: Here again, where do these value come from?

If this test were to fail, it wouldn't be clear if the test was wrong (and the new code was buggy) or if the test was correct and new version of the code introduced a regression. In the latter case, this test would not be very helpful to spot the root cause of the regression and that will make it tedious to debug.

It would be better to instead test on a very small data, e.g. with just 5 observations and put the expression of the computation of the expected as inline comments for each expected sample weight (with the KM formula).

File: hazardous/tests/test_ipcw.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+import pytest
+from lifelines.datasets import load_regression_dataset
+from numpy.testing import assert_array_almost_equal
+
+from .._ipcw import IpcwEstimator
+
+
+@pytest.mark.parametrize("competitive_risk", [True, False])
```
Comment (ogrisel) [inline hazardous/tests/test_ipcw.py:9]: `competitive` => `competing`.

This typo is repeated many times in this PR.

File: hazardous/tests/test_ipcw.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            1.19718884,
+            1.33020982,
+            1.33020982,
+            1.33020982,
+            1.33020982,
+            1.33020982,
+        ]
+    )
+
+    assert_array_almost_equal(ipcw_probs, expected_ipcw_probs)
```
Comment (ogrisel) [inline hazardous/tests/test_ipcw.py:52]: Instead of checking specific values that come from nowhere (but our current implementation), let's instead check expected mathematical properties:

```suggestion
    # Probabilities lie in [0.0, 1.0] hence the range of their inverse
    # is [1.0, +inf).
    assert (ipcw_probs >= 1.0).all(), ipcw_probs
    
    # Expect monotonically increasing censoring probabilities over time,
    # therefore monotonically increasing IPCW values.
    assert (ipcw_probs.diff() >= 0).all(), ipcw_probs
```

Also, nowadays, `assert_allclose` is considered a better replacement to `assert_array_almost_equal` (because the way tolerance is defined makes more sense usually).

Comment (ogrisel) [reply hazardous/tests/test_ipcw.py:52]: Actually I am confused now. Are IPCW "inverse probabilities of censoring" or "inverse probabilities of absence of censoring"?

File: hazardous/tests/test_ipcw.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+@pytest.mark.parametrize("competitive_risk", [True, False])
+def test_ipcw(competitive_risk):
+    X = load_regression_dataset()
+    y = dict(
+        event=X["E"],
+        duration=X["T"],
+    )
+    if competitive_risk:
+        coef = np.random.choice([1, 2], size=y["event"].shape[0])
```
Comment (ogrisel) [inline hazardous/tests/test_ipcw.py:17]: I think we should always seed RNGs in test to avoid random, rare, non-reproducible errors on the CI.

```suggestion
        rng = np.random.default_rng(0)
        coef = rng.choice([1, 2], size=y["event"].shape[0])
```

We could make the test non-rng sensitive by parameterizing it with a seed, e.g.:

```python
@pytest.mark.parametrize("competing_risks", [True, False])
@pytest.mark.parametrize("random_seed", range(3))
def test_ipcw(competing_risks, random_seed):
    ...
        rng = np.random.default_rng(random_seed)
        coef = rng.choice([1, 2], size=y["event"].shape[0])
    
```

File: hazardous/tests/test_metrics.py [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+            0.05829176,
+            0.0663998,
+            0.04524901,
+            0.04553689,
+            0.02250038,
+            0.02259133,
+        ]
+    )
+
+    assert_array_almost_equal(loss, loss_expected)
```
Comment (ogrisel) [inline hazardous/tests/test_metrics.py:272]: Similar comment as above: let's not test against meaning less hardcoded values but instead against:

- either invariant mathematical properties;
- hardcoded values that directly stem from minimal toy dataset where we can derive
  the  expected values analytically (and inline comments to put the derivation).

**Review by ogrisel — COMMENTED** [review]
First quick pass of feedback.

File: hazardous/tests/test_ipcw.py (hunk 5 above)
Comment (Vincent-Maladiere) [reply hazardous/tests/test_ipcw.py:52]: This values don't come from nowhere, they are the reference from scikit-survival and are far more reliable than some macro statistical checks.

Comment (Vincent-Maladiere) [reply hazardous/tests/test_ipcw.py:52]: IPCW = Inverse Probability of Censoring Weight (your first guess), we compute these with Kaplan Meier by considering censoring as the event of interest.

File: hazardous/tests/test_metrics.py (hunk 7 above)
Comment (Vincent-Maladiere) [reply hazardous/tests/test_metrics.py:272]: I agree with deriving stuff analytically, but having the results from scikit-survival allowed me to revamp the scorer and models faster. The issue is that the GPL license prevent me from importing it in the tests.

File: hazardous/tests/test_ipcw.py (hunk 5 above)
Comment (ogrisel) [reply hazardous/tests/test_ipcw.py:52]: > This values don't come from nowhere, they are the reference from scikit-survival and are far more reliable than some macro statistical checks.

Alright, then let's make this explicit in the code that makes it possible to reproduce those values with scikit-survival by adding the code snippet in commented out lines.

Comment (ogrisel) [reply hazardous/tests/test_ipcw.py:52]: > IPCW = Inverse Probability of Censoring Weight (your first guess), we compute these with Kaplan Meier by considering censoring as the event of interest.

Now I realize that this name (IPCW) is a bit misleading: 

KM on `(1 - any_event_indicator)` indeed estimates the **survival to censoring**. Hence, it estimates the cumulative probability of "not-censoring":

At t=0, the probability of censoring is always 0, and IPCW is typically at its lowest value, close to 1.0. When time progresses, the cumulative probability of censoring can only increases. The probability of its complement (probability of not censoring) monotically decreases and IPCW therefore monotonically increases (and always stays above 1.0 by construction).

On some datasets, the censoring probability reach 1.0 after t_max (all datapoints recorded with t > t_max are censored). In this case, the probability of not being censored is getting very close to 0 and IPCW can go to +inf.

I will try to make a suggestion to add a comment to make that explicit in the code.

File: hazardous/tests/test_metrics.py (hunk 7 above)
Comment (ogrisel) [reply hazardous/tests/test_metrics.py:272]: At least make it explicit that those values were computed with scikit-survival as a reference implementation:

https://github.com/soda-inria/hazardous/pull/10#discussion_r1257849411

**Comment by ogrisel** [comment]
Let's add some tests to check some mathematical properties:

- `IPCWEstimator` should constantly predict 1.0 when the training data has no censored values,
- check that `assert_allclose(ipcw.predict([0]), [1.0])` when `ipcw` is fit on random data, even when the shorter time in the training set is censored.
- check that `IPCWEstimator` predicts monotonically increasing values on a `np.linspace(0, t_max, 100)` when fit on random censored event data.
- check that when a training set has deterministic censoring beyond a given t value (100% of the largest training events are censored), predicting with large t values is clipped to `1 / ipcw.min_censoring_survival_prob` for different values of `min_censoring_survival_prob`.

File: hazardous/tests/test_metrics.py [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        shape=y["event"].shape[0],
+        dtype=[("event", np.int32), ("duration", np.float64)],
+    )
+    y_out["event"] = y["event"]
+    y_out["duration"] = y["duration"]
+    return y_out
+
+
+@pytest.mark.parametrize("format_func", [_dict_to_pd, _dict_to_recarray])
+def test_inputs_format(format_func):
```
Comment (ogrisel) [inline hazardous/tests/test_metrics.py:241]:
```suggestion
def test_brier_score_inputs_format(format_func):
```

File: hazardous/tests/test_metrics.py [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+            0.04553689,
+            0.02250038,
+            0.02259133,
+        ]
+    )
+
+    assert_array_almost_equal(loss, loss_expected)
+
+
+def test_wrong_inputs():
```
Comment (ogrisel) [inline hazardous/tests/test_metrics.py:275]:
```suggestion
def test_brier_score_wrong_inputs():
```

File: hazardous/tests/test_metrics.py [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+            1.05622555,
+            1.0892326,
+            1.03891038,
+        ]
+    )
+
+    assert_array_almost_equal(sample_weights[:10], sample_weights_expected)
+
+
+def test_wrong_parameters():
```
Comment (ogrisel) [inline hazardous/tests/test_metrics.py:202]:
```suggestion
def test_brier_score_wrong_parameters():
```

File: hazardous/metrics/brier_score.py [hunk 11]
```diff
@@ -0,0 +1,471 @@
...
+        y_train,
+        y_test,
+        y_pred,
+        times,
+        event_of_interest=event_of_interest,
+    )
+    return np.trapz(brier_scores, times) / (times[-1] - times[0])
+
+
+class BrierScoreSampler(BrierScoreComputer):
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:439]: I think this should not be part of the `hazardous.metrics` module, but instead could be moved into the source file of the estimator that uses it.

Same remark for the test.

File: hazardous/metrics/brier_score.py [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+    """Sample random times uniformly to compute the IPCW.
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:440]:
```suggestion
    """Sample random times uniformly to compute the time-dependent Brier score.
```

File: hazardous/metrics/brier_score.py [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+
+
+def brier_score_incidence(
+    y_train,
+    y_test,
+    y_pred,
+    times,
+    event_of_interest="any",
+):
+    """Compute the Brier Score for the kth cause of failure.
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:313]:
```suggestion
    """Compute the time dependent Brier score for the kth cause of failure.
```

File: hazardous/metrics/brier_score.py [hunk 14]
```diff
@@ -0,0 +1,471 @@
...
+            In the binary event settings, this is 1 - incidence_probability.
+
+        times : array-like of shape (n_times)
+            Times to estimate the survival probability and to compute the
+            Brier Score.
+
+        Returns
+        -------
+        brier_score : np.ndarray
+            Average value of individual Brier Scores computed for ``times``.
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:75]:
```suggestion
            Time-dependent Brier scores averaged over the individuals.
```

File: hazardous/metrics/brier_score.py [hunk 15]
```diff
@@ -0,0 +1,471 @@
...
+
+        # Precompute the censoring probabilities at the time of the events on the
+        # training set:
+        self.ipcw_train = self.ipcw_est.predict(self.duration_train)
+
+    def brier_score(self, y_true, y_pred, times):
+        """Compute the Brier Score.
+
+        For each sample, apply the Brier Score formula, then
+        average each individual Brier Score column-wise.
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:57]:
```suggestion
        """Time-dependent Brier score of a survival function estimate.

        Compute the time-dependent Brier score value for each individual and each
        time point in `times` and then average over individuals.

        This estimate is adjusted for censoring by leveraging the Inverse Probability
        of Censoring Weighting (IPCW) scheme.
```

**Review by ogrisel — COMMENTED** [review]
More feedback.

File: hazardous/metrics/brier_score.py [hunk 16]
```diff
@@ -0,0 +1,471 @@
...
+        \\mathrm{BS}(t) = \\frac{1}{n} \\sum_{i=1}^n \\mathbb{I}
+        (y_i \\leq t \\land \\delta_i = 1)
+        \\frac{(0 - \\hat{S}(t | \\mathbf{x}_i))^2}{\\hat{G}(y_i)} +
+        \\mathbb{I}(y_i > t)
+        \\frac{(1 - \\hat{S}(t | \\mathbf{x}_i))^2}{\\hat{G}(t)} ,
+
+    where :math:`\\hat{S}(t | \\mathbf{x})` is the predicted probability of
+    surviving up to time point :math:`t` for a feature vector :math:`\\mathbf{x}`,
+    and :math:`1/\\hat{G}(t)` is a inverse probability of censoring weight, estimated by
+    the Kaplan-Meier estimator.
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:218]:
```suggestion
    and :math:`\\hat{G}(t)` is the probability of remaining uncensored at time
    :math:`t`, estimated on the training set by the Kaplan-Meier estimator on the
    negation of the binary any-event indicator.
```

File: hazardous/metrics/brier_score.py [hunk 17]
```diff
@@ -0,0 +1,471 @@
...
+
+    Parameters
+    ----------
+    y_train : record-array, dictionnary or dataframe of shape (n_samples, 2)
+        The target, consisting in the 'event' and 'duration' columns.
+        This is used to fit the IPCW estimator.
```
Comment (ogrisel) [inline hazardous/metrics/brier_score.py:224]:
```suggestion
        This is used to estimate the IPCW values to adjust for censoring in the
        evaluation data.
```

File: pyproject.toml [hunk 18]
```diff
@@ -50,7 +51,7 @@ doc = [
 
 [build-system]
 build-backend = "flit_core.buildapi"
-requires = ["flit_core"]
+requires = ["flit_core", "setuptools"]  # setuptools is needed to install pkg_resources, used in lifelines
```
Comment (ogrisel) [inline pyproject.toml:54]: It's weird to have to update a the hazardous build dependencies this way. When do you get an error if `setuptools` is not installed? At build time (when building the hazardous wheel from source) or at runtime (when importing or using hazardous)?

If at runtime, we should instead add setuptools as a runtime dependency (and open an issue upstream in lifelines to let them know that they have an implicit runtime dependency on setuptools that should be declared as such in their package meta-data.

File: hazardous/utils.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        raise ValueError(
+            "event_of_interest must be a strictly positive integer or 'any', "
+            f"got: event_of_interest={k}"
+        )
+    return
+
+
+def check_y_mean_increasing(y_pred, times):
+    """Check that the mean of y is increasing
+    when sorting with times.
```
Comment (ogrisel) [inline hazardous/utils.py:48]:
```suggestion
    """Check that the mean of y is increasing with time.
```

File: hazardous/utils.py [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+
+    This allow to warn users of spurious survival probability inputs,
+    when the incidence probability is expected instead.
```
Comment (ogrisel) [inline hazardous/utils.py:51]:
```suggestion
    This allows to warn users of spurious survival probability inputs,
    when the incidence probability is expected instead.
```

File: hazardous/tests/test_ipcw.py (hunk 5 above)
Comment (ogrisel) [reply hazardous/tests/test_ipcw.py]: I updated the code, docstring and tests to take those remarks into account.

**Comment by ogrisel** [comment]
Ok @Vincent-Maladiere I think this is ready for final pass of review of your end.

**Comment by ogrisel** [comment]
I rendered the HTML of the doc locally but I realize that we do not have a CI job to do it for the PR. Only for publishing the result in the merge of the PR.
